doc/cpp-style-guide.md

```diff
@@ -0,0 +1,933 @@
+getting used to, but it is important that all project contributors follow the style rules so that they can all read
+and understand everyone's code easily.
+
+The project enforces formatting using `clang-format`, configured to follow the `Google` style guide.  Please
+configure your editor or IDE to follow the same guidelines.  When the tool disagrees with this document please file a
+bug either the tool configuration or the document needs adjustment.
+
+In a few cases where the GSG give latitude we have made a choice.
+
+***Line length:***
+
+Each line of text in your code should be at most 120 characters long.
+
+#### Non-ASCII Characters
+
+Non-ASCII characters should be rare, and must use UTF-8 formatting.
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#Non-ASCII_Characters)
+
+#### Spaces vs. Tabs
+
+Use only spaces, and indent 2 spaces at a time.
+
+We use spaces for indentation. Do not use tabs in your code. You should set your editor to emit spaces when you hit
+the tab key.
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#Spaces_vs._Tabs)
+
+#### Function Declarations and Definitions
+
+Return type on the same line as function name, parameters on the same line if they fit. Wrap parameter lists which do not fit on a single line as you would wrap arguments in a function call.
+
+https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions
+#### Lambda Expressions
+
+Format parameters and bodies as for any other function, and capture lists like other comma-separated lists.
+
+For by-reference captures, do not leave a space between the ampersand (&) and the variable name.
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#Formatting_Lambda_Expressions)
+
+#### Function Calls
+
+Either write the call all on a single line, wrap the arguments at the parenthesis, or start the arguments on a new line indented by four spaces and continue at that 4 space indent. In the absence of other considerations, use the minimum number of lines, including placing multiple arguments on each line where appropriate.
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#Function_Calls)
+
+#### Braced Initializer List Format
+
+Format a braced initializer list exactly like you would format a function call in its place.
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#Braced_Initializer_List_Format)
+
+#### Conditionals
+
+Prefer no spaces inside parentheses. The if and else keywords belong on separate lines.
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#Conditionals)
+
+#### Loops and Switch Statements
+
+Switch statements may use braces for blocks. Annotate non-trivial fall-through between cases. Braces are optional for single-statement loops. Empty loop bodies should use empty braces or continue.
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#Loops_and_Switch_Statements)
+
+#### Pointer and Reference Expressions
```

Comment:
Can we collapse all formatting rules under a simple rule - "All code checked in must be run through the provided .clang-format style. Turning off clang-format can be used sparingly for sections of a file where it improves readability".  Would be nice if we can condition developers of this project to not be attached to their indentation style and rather defer to whatever clang-format does on the given file.

Reply:
Well, we already use `clang-format` in the CI builds and your build breaks if you do not follow the formatting.  The question this document is trying to answer is "How should we configure .clang-format"?  And basically we are saying:  `BasedOnStyle: Google` with some tweaks (the tweaks are in this PR too).

Reply:
The .clang-format spec is a much terser way of specifying the same rules and maybe we should just point users to refer to that for formatting guidelines if they are curious about it. I am hoping that majority of developers will never need to learn about the formatting rules and will just use clang-format. Dropping formatting rules from this document will leave the remaining document focused and likely to fit in our users L1 cache.

Reply:
I think we are in agreement, I basically already say this in the CONTRIBUTING.md doc  [link](https://github.com/GoogleCloudPlatform/google-cloud-cpp/blob/master/CONTRIBUTING.md#formatting), and earlier in this document too.  I am adding some clarification as to why we repeat the formatting rules (which is basically "to compare against the GSG").